### PR TITLE
Dovecot: make home dir distinct from mail dir

### LINKED ISSUE #3296: [BUG]: Dovecot mail-dir and home locations shouldn't overlap
### Preliminary Checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.
- [X] I searched the issue tracker but was unable to find my issue.
- [X] I read the [extended documentation in general](https://docker-mailserver.github.io/docker-mailserver/latest/) but found nothing to resolve the issue.
- [X] I read the [documentation on debugging](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/), tried the proposed steps to debug the problem, but was still unable to resolve the issue.
- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree

### Affected Component(s)

Dovecot errors when creating a filter rule to forward an email

### What happened and when does this occur?

Logwatch reports the following (sanitized) error:

imap([<EMAIL_ADDRESS>](mailto:<EMAIL_ADDRESS>))<1492788><5ivHyk/6QQSsOq3S>: Error: stat(/var/mail/<DOMAIN>/<USER>/.dovecot.lda-dupes/tmp) failed: Not a directory: 2 Time(s)

### What did you expect to happen?

Clean log files in logwatch

### How do we replicate the issue?

1.Using Roundcube I created a rule to forward specific emails from one local mail user to another mail user (error seems to have started when I did this)
2. This creates a file named .dovecot.lda-dupes in the users maildir folder
3. Dovecot appears to think this is an email folder because of it's location and because it starts with a dot
...

### DMS version

v12.0

### What operating system is DMS running on?

Linux

### Which operating system version?

debian 11

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker

### docker-compose.yml

```yml
services:
  mailserver:
    image: docker.io/mailserver/docker-mailserver:latest
    container_name: mailserver
    # If the FQDN for your mail-server is only two labels (eg: example.com),
    # you can assign this entirely to `hostname` and remove `domainname`.
    hostname: mail
    domainname: edmunds.email
    env_file: mailserver.env
    #dns:
    #  - 127.0.0.1
    # More information about the mail-server ports:
    # https://docker-mailserver.github.io/docker-mailserver/edge/config/security/understanding-the-ports/
    # To avoid conflicts with yaml base-60 float, DO NOT remove the quotation marks.
    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS)
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
      - "4190:4190"    # Manage sieve
      - "10110:10110"  # POP3 (explicit TLS with HAProxy)
      - "10123:10123"  # Dovecot auth port for radicale
      - "10143:10143"  # IMAP4 (explicit TLS with HAProxy)
      - "10993:10993"  # IMAP4 (implicit TLS with HAProxy)
      - "10995:10995"  # POP3S (implicit TLS with HAProxy)
      - "11334:11334"  # Rspamd web interface
      - "14190:14190"  # Manage sieve
    volumes:
      - ./docker-data/dms/mail-data/:/var/mail/
      - ./docker-data/dms/mail-state/:/var/mail-state/
      - ./docker-data/dms/mail-logs/:/var/log/mail/
      - ./docker-data/dms/config/:/tmp/docker-mailserver/
      - /etc/localtime:/etc/localtime:ro
      - ./docker-data/dms/certs/:/tmp/dms/certs/:ro
      # ./docker-data/dms/cron/sa-learn:/etc/cron.d/sa-learn
      - ./docker-data/dms/config/dovecot/10-custom.conf:/etc/dovecot/conf.d/10-custom.conf
      - ./docker-data/dms/config/rspamd/:/etc/rspamd/override.d/
    restart: unless-stopped
    stop_grace_period: 1m
    cap_add:
      - NET_ADMIN
    healthcheck:
      test: "ss --listening --tcp | grep -P 'LISTEN.+:smtp' || exit 1"
      timeout: 3s
      retries: 0
```

### Relevant log output

```Text
imap([<EMAIL_ADDRESS>](mailto:<EMAIL_ADDRESS>))<1492788><5ivHyk/6QQSsOq3S>: Error: stat(/var/mail/<DOMAIN>/<USER>/.dovecot.lda-dupes/tmp) failed: Not a directory: 2 Time(s)
```

### Other relevant information

According to this post the location of this file is an issue:
https://dovecot.dovecot.narkive.com/BRuWMv1g/problem-with-lda-dupes-file

```
Because it starts with a dot and is located where your Maildir-root
directory is and the IMAP demon tries to access a mailbox called
"dovecot.lda-dupes".

You should not (must not) have home == maildir root.
```

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [ ] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [ ] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [ ] I am rather comfortable with the CLI

### Improvements to this form?

_No response_

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

Funnily enough, strictly speaking this is not actually a breaking change. The Sieve scripts are already moved by us from `/tmp/docker-mailserver` to the correct location on startup, and the same goes for the Dovecot userdb file. But that is irrelevant, as the next version we are going to release is a major.. I woudl definitely mention it separately on the release page and in the changelog.

I consider this PR as high priority, as <<https://doc.dovecot.org/configuration_manual/home_directories_for_virtual_users/> points out: "Home directory shouldn’t be the same as mail directory with mbox or Maildir formats (but with dbox/obox it’s fine). It’s possible to do that, but you might run into trouble with it sooner or later."

I took care that mdbox/sdbox formats are untouched.

Fixes #3296

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by polarathene** [comment]
> I took care that mdbox/sdbox formats are untouched.

Not sure if that's required? The docs seem to imply that it's a non-issue for them, but nothing about having a separate location for the user home like other mailbox types.

> I would definitely mention it separately on the release page and in the changelog.

Presumably this will be a problem for users unaffected as separating out the content for N users to the new home subdir may not be straight-forward?

I'm unable to investigate if that is a possibility to advise further. It may warrant supporting a way to opt-in to keeping the current home location, but I leave that decision up to you.

**Review by polarathene — DISMISSED** [review]

**Comment by georglauterbach** [comment]
> > I took care that mdbox/sdbox formats are untouched.
> 
> Not sure if that's required? The docs seem to imply that it's a non-issue for them, but nothing about having a separate location for the user home like other mailbox types.

I guess this is because SDBOX/MDBOX seem to already use different directories for that by default. When looking at the tests, there is the "`mailboxes`" directory, which already does the separation AFAIK. Hence, we needn't update it.

> > I would definitely mention it separately on the release page and in the changelog.
> 
> Presumably this will be a problem for users unaffected as separating out the content for N users to the new home subdir may not be straight-forward?

I'm not sure I understand. Those that use MDBOX/SDBOX already use a different directory structure altogether. What is important is to get everyone in-line when the switch to DMS >= v13.0.0 (i.e. everyone who uses `maildir` must get the new home directory).

> I'm unable to investigate if that is a possibility to advise further. It may warrant supporting a way to opt-in to keeping the current home location, but I leave that decision up to you.

I'd like to avoid that, due to what I just said (above). I don't want to maintain partitions of users with one home dir, and the other partition with another home dir..

**Comment by casperklein** [comment]
It's not completely clear to me. At the moment, `home dir == mail dir`. What files are supposed to be in home dir?

Is it fine to just use the new configuration with home/mail dirs separated? Or is there some kind of data migration needed?

This is how a directory looks at the moment on my system:

```
root@mail:/var/mail/example.com/postmaster# ls
total 252K
drwxr-xr-x 2 docker docker  44K May 14 16:44 cur
-rw-r--r-- 1 docker docker  13K May 14 17:54 dovecot-uidlist
-rw-r--r-- 1 docker docker    8 Mar  4  2020 dovecot-uidvalidity
-r--r--r-- 1 docker docker    0 Mar  4  2020 dovecot-uidvalidity.5e6032a8
-rw-r--r-- 1 docker docker 2.9K May 14 06:36 dovecot.index
-rw-r--r-- 1 docker docker  80K May 14 17:54 dovecot.index.cache
-rw-r--r-- 1 docker docker 7.6K May 14 17:54 dovecot.index.log
-rw-r--r-- 1 docker docker  33K May 14 06:36 dovecot.index.log.2
-rw-r--r-- 1 docker docker  384 Mar  4  2020 dovecot.list.index.log
drwxr-xr-x 2 docker docker  36K May 14 17:54 new
drwxr-xr-x 2 docker docker 4.0K May 14 17:54 tmp
```

What files will be in `/var/mail/example.com/postmaster/home` after his change?

**Comment by georglauterbach** [comment]
At the moment, this is primarily about Sieve user files. From what I can see, nothing from your directory @casperklein is relevant when it comes to the home dir.

> Is it fine to just use the new configuration with home/mail dirs separated? Or is there some kind of data migration needed?

AFAIK it actually is. This is mainly due to the reason the mail dir did not change, only the home dir location did.

---

Here is an excerpt from the Dovecot docs:

> Some uses for home directory are:
>
> - By default [Sieve](https://wiki.dovecot.org/Pigeonhole/Sieve) scripts are in user's home directory.
> - Duplicate mail check database is in user's home directory. Suppression of duplicate rejects/vacations won't work if home directory isn't specified.
> - Debugging: If an imap or pop3 process crashes, the core file is written to the user's home directory.

File: target/scripts/startup/setup.d/dovecot.sh [hunk 1]
```diff
@@ -24,18 +24,17 @@ function _setup_dovecot
...
         /etc/dovecot/conf.d/10-mail.conf
-
       _log 'trace' 'Enabling cron job for dbox purge'
       mv /etc/cron.d/dovecot-purge.disabled /etc/cron.d/dovecot-purge
       chmod 644 /etc/cron.d/dovecot-purge
       ;;
 
     ( * )
       _log 'trace' 'Dovecot default format (maildir) configured'
-      sed -i -e 's|^mail_location = .*$|mail_location = maildir:\/var\/mail\/%d\/%n|g' /etc/dovecot/conf.d/10-mail.conf
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/scripts/startup/setup.d/dovecot.sh:38]: Exactly!

File: target/dovecot/10-mail.conf [hunk 2]
```diff
@@ -27,6 +27,7 @@
 #
 # <doc/wiki/MailLocation.txt>
 #
+mail_home = /var/mail/%d/%n/home
```
[the remark this answers is not in the document]
Comment (georglauterbach) [reply target/dovecot/10-mail.conf:30]: Agreed :)

File: target/dovecot/auth-passwdfile.inc [hunk 3]
```diff
@@ -15,5 +15,5 @@ passdb {
 userdb {
   driver = passwd-file
   args = username_format=%u /etc/dovecot/userdb
-  default_fields = uid=docker gid=docker home=/var/mail/%d/%u
+  default_fields = uid=docker gid=docker home=/var/mail/%d/%u/home
```
Comment (casperklein) [inline target/dovecot/auth-passwdfile.inc:18]:
```suggestion
  default_fields = uid=docker gid=docker home=/var/mail/%d/%u/home/
```

**Review by casperklein — DISMISSED** [review]
Another trailing slash.. otherwise ready to merge.

**Comment by georglauterbach** [comment]
I will merge this and then update my system to check whether everything is fine.

**Review by casperklein — APPROVED** [review]

**Comment by alterNERDtive** [comment]
So just to be perfectly clear now that 13.0.0 is released: unless I have fiddled with Dovecot settings like a madman, I won’t have to manually touch anything even though this is listed as a breaking change in the release notes?

**Comment by Sovamorco** [comment]
After 13.0.0 was pushed as latest, at least for me - all Sieve scripts stopped working, and managesieve was showing none of them. Had to manually move all scripts from `server/account/sieve` to `server/account/home/sieve` and re-enable them with managesieve

**Comment by georglauterbach** [comment]
I will make sure to mention this in the release notes; sorry for not making this more clear in `CHANGELOG.md`.

**Pull request merged**